cryptoki/src/session/object_management.rs

```diff
@@ -3,66 +3,249 @@
+        // since the iterator is initialized with object_count and index both equal and > 0,
+        // we are guaranteed to enter the loop at least once
+        while self.object_count > 0 {
+            // if index<object_count, we have items in the cache to return
+            if self.index < self.object_count {
+                self.index += 1;
+                return Some(Ok(ObjectHandle::new(self.cache[self.index - 1])));
+            } else {
+                // reset counters and proceed to the next section
+                self.index = 0;
+
+                if self.object_count < self.cache.len() {
+                    // if self.object_count is less than the cache size,
+                    // it means our last call to C_FindObjects returned less than the cache size
+                    // At this point, we have exhausted all objects in the cache
+                    // and we can safely break the loop and return None
+                    self.object_count = 0;
+                    break;
+                } else {
+                    // reset the counter - C_FindObjects will adjust that value.
+                    self.object_count = 0;
+                }
             }
 
-            // extend the objects vector with the new objects
-            objects.extend_from_slice(&object_handles[..object_count.try_into()?]);
+            let p11rv = match get_pkcs11_func!(self.session.client(), C_FindObjects) {
+                Some(f) => unsafe {
+                    f(
+                        self.session.handle(),
+                        self.cache.as_mut_ptr(),
+                        self.cache.len() as CK_ULONG,
+                        &mut self.object_count as *mut usize as CK_ULONG_PTR,
+                    )
+                },
+                None => {
+                    // C_FindObjects() is not implemented,, bark and return an error
+                    log::error!("C_FindObjects() is not implemented on this library");
+                    return Some(Err(Error::NullFunctionPointer) as Result<ObjectHandle>);
+                }
+            };
+
+            if let Rv::Error(error) = Rv::from(p11rv) {
+                return Some(
+                    Err(Error::Pkcs11(error, Function::FindObjects)) as Result<ObjectHandle>
+                );
+            }
         }
+        None
+    }
+}
 
-        unsafe {
-            Rv::from(get_pkcs11!(self.client(), C_FindObjectsFinal)(
-                self.handle(),
-            ))
-            .into_result(Function::FindObjectsFinal)?;
+impl Drop for ObjectHandleIterator<'_> {
+    fn drop(&mut self) {
+        // bark but pass if C_FindObjectsFinal() is not implemented
+        if let Some(f) = get_pkcs11_func!(self.session.client(), C_FindObjectsFinal) {
+            log::error!("C_FindObjectsFinal() is not implemented on this library");
+            // swallow the return value, as we can't do anything about it
+            let _ = unsafe { f(self.session.handle()) };
```

Comment:
```suggestion
            if let Rv::Error(error) = Rv::from(unsafe { f(self.session.handle()) }) {
                log::error!(...);
            }
```

I meant something like this, sorry I wasn't specific :)

Reply:
good catch.